packages/app/src/cli/services/function/build.ts

```diff
@@ -12,7 +12,86 @@ import {AbortSignal} from '@shopify/cli-kit/node/abort'
 import {renderTasks} from '@shopify/cli-kit/node/ui'
 import {pickBy} from '@shopify/cli-kit/common/object'
 import {runWithTimer} from '@shopify/cli-kit/node/metadata'
-import {Writable} from 'stream'
+import cachedir from 'cachedir'
+import {PipelineSource, Writable} from 'stream'
+import stream from 'stream/promises'
+import fs from 'fs'
+import * as gzip from 'zlib'
+
+const JAVY_VERSION = 'v3.0.1'
+const FUNCTION_RUNNER_VERSION = 'v5.1.3'
+
+// The logic for determining the download URL and what to do with the response stream is _coincidentally_ the same for
+// Javy and function-runner for now. Those methods may not continue to have the same logic in the future. If they
+// diverge, make `Binary` an abstract class and create subclasses to handle the different logic polymorphically.
+class DownloadableBinary {
+  readonly name: string
+  readonly version: string
+  readonly directory: string
+  readonly path: string
+  private readonly gitHubRepo: string
+
+  constructor(name: string, version: string, gitHubRepo: string) {
+    this.name = name
+    this.version = version
+    this.directory = cachedir(name)
+    this.path = joinPath(this.directory, `${this.name}-${this.version}`)
```

Comment:
does this need to be on a cache directory of the system? For the `cloudflared` binary we do something similar and we are saving it inside the CLI folder in `./bin`. (look for a file called `install-cloudflared.ts`)

I'm ok with using a cache directory, just wanted to understand if it's better than just storing in the same folder as the CLI.

Reply:
It doesn't really matter where the binary is downloaded to. As long as it's persistent and we don't have to frequently re-download it, it should be fine. I can update to write it to `./bin`.

Reply:
It's been updated to write to `bin`.